lib/cretonne/meta/cdsl/registers.py

```diff
@@ -195,24 +195,29 @@ class RegClass(object):
     :param start: The first unit to allocate, relative to `bank.first.unit`.
     """
 
-    def __init__(self, bank, count=None, width=1, start=0):
-        # type: (RegBank, int, int, int) -> None
+    def __init__(self, bank, count=0, width=1, start=0, bitmask=None):
+        # type: (RegBank, int, int, int, int) -> None
         self.name = None  # type: str
         self.index = None  # type: int
         self.bank = bank
-        self.start = start
         self.width = width
+        self.bitmask = 0
 
         # This is computed later in `finish_regclasses()`.
         self.subclasses = list()  # type: List[RegClass]
         self.toprc = None  # type: RegClass
 
         assert width > 0
-        assert start >= 0 and start < bank.units
 
-        if count is None:
-            count = bank.units // width
-        self.count = count
+        if bitmask:
+            self.bitmask = bitmask
+        else:
+            assert start >= 0 and start < bank.units
+            if count == 0:
+                count = bank.units // width
+            for a in range(count):
+                u = start + a * self.width
+                self.bitmask |= 1 << u
```

Comment:
Intel register classes all have width 1, but ARM has D-registers that each take up one S-register and Q-registers that each take up 4 S-registers. This is modeled with width=2 and width=4 respectively.

When width > 1, I think the bit mask should have a set bit for all of the units covered by the register, rather than just the first one.

You can do that without a loop like this: `(1 << (count*self.width)) - 1) << start`.

Reply:
This doesn't seem to match the way that RegClass masks are currently generated. For instance:

```
--- good/registers-arm32.rs	2018-01-16 18:35:54.000000000 -0800
+++ bad/registers-arm32.rs	2018-01-16 18:36:15.000000000 -0800
@@ -60,7 +60,7 @@
     toprc: 1,
     first: 0,
     subclasses: 0x2,
-    mask: [0x55555555, 0x55555555, 0x00000000],
+    mask: [0xffffffff, 0xffffffff, 0x00000000],
     info: &INFO,
 };
 #[allow(dead_code)]
@@ -73,7 +73,7 @@
     toprc: 2,
     first: 0,
     subclasses: 0x4,
-    mask: [0x11111111, 0x11111111, 0x00000000],
+    mask: [0xffffffff, 0xffffffff, 0x00000000],
     info: &INFO,
 };
 #[allow(dead_code)]
```

This diff is of the generated registers-arm32.rs file, the former being how it works in master and the latter with this change. The top part of the diff is the `D` register class, and the bottom is `Q`.

This change also causes the `pressure.rs` unit tests to fail.

Reply:
My mistake. Sorry about that!